numpy/core/src/multiarray/common.c

```diff
@@ -441,11 +441,16 @@ PyArray_DTypeFromObjectHelper(PyObject *obj, int maxdims,
     }
 
     /* Recursive case, first check the sequence contains only one type */
+    /*
+     * The C-API recommends calling PySequence_Fast before any of the other
+     * PySequence_Fast* functions. This is required for PyPy
+     */
```

Comment:
I think you might want to move this comment and the PySequence_Fast call to before the `/* Recursive case ... */` comment

Reply:
done